### PR TITLE
Font lock types in their own color

### LINKED ISSUE #899: Font-lock types differently
Types and constructors share namespace in Haskell and only context tell how a token starting with uppercase letter should be interpreted. Emacs' font lock isn't best at context but can be coerced to use one. The plan:

1. Trigger type font lock on `::` token.
2. Parse what comes after that as a type expression.
3. Mark where the type expression starts and ends.
4. Font lock tokens in marked region as types instead of constructors.

Tricky situations:

```haskell
putStrLn (123 :: Int)
x :: String <- return ""
x :: String = "xx"
x :: (HasField "field" r) => r
x + y :: Int where y = 45
data T = C
```

And everything related to `DataKinds`.

### PR DESCRIPTION
**Pull request by gracjan**
Fixes #899.

### CODE REVIEW

**Comment by gracjan** [comment]
So it currently looks like this:

font-lock-types

**Comment by gracjan** [comment]
Todo:

- [ ] multiline cases do not refresh properly, possibly use `font-lock-extend-region-functions`
- [ ] handle `class` and `instance` with functional deps
- [ ] handle `data` ... `where` (looks easy)
- [ ] handle normal `data X1 = X2 Int String` (this is hard, has to know that X2 is constructor but Int and String are types )
- [ ] write lots of unit cases
- [ ] handle `module (...)` header (very hard)
- [ ] patterns? views?
- [ ] handle `type MyInt = Int` (hard, has to cross equal sign)
- [ ] handle `data family` and `type family`

**Comment by gracjan** [comment]
In case of `data X = X1 Int | X2 String` everything is a type except uppercase constructors directly following equals or vertical bar.

**Comment by geraldus** [comment]
Nice solution (:

**Comment by gracjan** [comment]
So both patterns and views work only with constructors.

File: haskell-mode.el [hunk 1]
```diff
@@ -501,6 +501,63 @@ executable found in PATH.")
...
+      (cond
+       ((null token)
+        (setq cont nil))
+       ((member token '(newline))
+        (setq last-token-was-newline (not last-token-was-operator))
+        (goto-char (match-end 0)))
+       ((and (member (match-string-no-properties 0)
+                     '("<-" "=" "<-" "←"  "," ";"
+                       ")" "]" "}" "|"
+                       "case" "class" "data" "default" "deriving" "do"
```
Comment (geraldus) [inline haskell-mode.el]: can `haskell-font-lock--reverved-ids` be re-used here? (:

Comment (gracjan) [inline haskell-mode.el]: Sorta maybe. I mean I'm to tired to make sure that all keywords end types. Is `forall` a keyword? is there other keyword that can happen in types?

Comment (gracjan) [inline haskell-mode.el]: And, btw, thanks for reading this. It wasn't easy to write so I'm grateful for all the help.

Comment (gracjan) [inline haskell-mode.el]: Fixed.

**Comment by gracjan** [comment]
Mostly done. Lets give it a couple of days lying here and then we merge.

File: haskell-mode.el (hunk 1 above)
Comment (geraldus) [inline haskell-mode.el]: I always read everything, PRs, issues, and code (: But, unfortunately most of the time I have nothing to add ):

**Pull request merged**